### PR TITLE
Benchmarks:  micro benchmark - source code for evaluating NVDEC decoding performance

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
source code for evaluating NVDEC decoding performance.

### CODE REVIEW

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    // With PreferCUVID, JPEG is still decoded by CUDA while video is decoded by NVDEC hardware
+    videoDecodeCreateInfo.ulCreationFlags = cudaVideoCreate_PreferCUVID;
+    videoDecodeCreateInfo.ulNumDecodeSurfaces = nDecodeSurface;
+    videoDecodeCreateInfo.vidLock = m_ctxLock;
+    videoDecodeCreateInfo.ulWidth = pVideoFormat->coded_width;
+    videoDecodeCreateInfo.ulHeight = pVideoFormat->coded_height;
+    // AV1 has max width/height of sequence in sequence header
+    if (pVideoFormat->codec == cudaVideoCodec_AV1 && pVideoFormat->seqhdr_data_length > 0) {
+        // dont overwrite if it is already set from cmdline or reconfig.txt
+        if (!(m_nMaxWidth > pVideoFormat->coded_width || m_nMaxHeight > pVideoFormat->coded_height)) {
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h:210]: Why not check m_nMaxWidth and m_nMaxHeight separately? It's strange.

Comment (cp5555) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h:210]: Also, it's not related to "// dont overwrite if it is already set from cmdline or reconfig.txt"

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+        *pnFrame = nFrame;
+    } catch (std::exception &) {
+        ex = std::current_exception();
+    }
+}
+
+/**
+ *   @brief  Function to show help message and exit
+ */
+void ShowHelpAndExit(const char *szBadOption = NULL) {
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp:62]: How can I get the default setting?

**Review by cp5555 — APPROVED** [review]

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+        // Calculate and output the raw data into file and metrics into stdout
+        double sum, mean, min, max, p50, p90, p95, p99;
+        std::tie(sum, mean, min, max, p50, p90, p95, p99) = CalLatencyMetrics(vnLatency);
+        if (outputFilePath.size() != 0) {
+            WriteRawData(vnLatency, vnFrame, outputFilePath);
+        }
+        std::cout << "Total Frames Decoded=" << nTotalFrames << " FPS=" << nTotalFrames / elapsedTime
+                  << " LatencyPerFrame=" << elapsedTime / nTotalFrames * 1000
+                  << " Mean Latency for each video=" << mean * 1000 << " P50 Latency=" << p50 * 1000
+                  << " P90 Latency=" << p90 * 1000 << " P95 Latency=" << p95 * 1000 << " P99 Latency=" << p99 * 1000
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp:409]: I think we can output two-layer metrics: Hardware level (FPS and Frame Latency) and Data level (VideoFPS and VideoLatency)

File: third_party/Video_Codec_SDK/Samples/Utils/Logger.h [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+private:
+    LogLevel level;
+    char szLead[80];
+    bool bPrintTimeStamp;
+    std::mutex mtx;
+};
+
+class LoggerFactory {
+public:
+    static Logger* CreateFileLogger(std::string strFilePath,
```
Comment (hippogr) [inline third_party/Video_Codec_SDK/Samples/Utils/Logger.h:101]: >(std::string [](http://example.com/codeflow?start=35&length=12)

use reference?

File: third_party/Video_Codec_SDK/Samples/Utils/Logger.h [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+private:
+    LoggerFactory() {}
+
+    class FileLogger : public Logger {
+    public:
+        FileLogger(std::string strFilePath, LogLevel level, bool bPrintTimeStamp) 
+        : Logger(level, bPrintTimeStamp) {
+            pFileOut = new std::ofstream();
+            pFileOut->open(strFilePath.c_str());
+        }
```
Comment (hippogr) [inline third_party/Video_Codec_SDK/Samples/Utils/Logger.h:122]: I guess there may have exception here, so please check it to make sure this call succeeds

File: third_party/Video_Codec_SDK/Samples/Utils/Logger.h [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+                    return;
+                }
+#ifdef _WIN32
+                closesocket(socket);
+                WSACleanup();
+#else
+                close(socket);
+#endif
+            }
+            void Flush() {
```
Comment (hippogr) [inline third_party/Video_Codec_SDK/Samples/Utils/Logger.h:183]: >Flush [](http://example.com/codeflow?start=17&length=5)

question: if the constructor fails to create the socket, what happens when you call Flush()? Will it cause the crash?

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+     * @brief This function is used to Get the frameLatency vector
+     */
+    std::vector<std::tuple<int, double>> &GetFrameLatency() { return frameLatency; }
+
+  protected:
+    /**
+     *   @brief  Callback function to be registered for getting a callback when decoding of sequence starts
+     */
+    static int CUDAAPI HandleVideoSequenceProc(void *pUserData, CUVIDEOFORMAT *pVideoFormat) {
+        return ((OptimizedNvDecoder *)pUserData)->HandleVideoSequence(pVideoFormat);
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h:37]: >pUserData [](http://example.com/codeflow?start=38&length=9)

do we need to check if the pointer is valid or not?

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+     *           Other part is the same as the original function, refer to NvDecoder.cpp in NVIDIA Video Codec SDK.
+     */
+    int HandleVideoSequence(CUVIDEOFORMAT *pVideoFormat);
+
+    CUVIDDECODECAPS m_decodecaps;
+
+    std::vector<std::tuple<int, double>> frameLatency;
+};
+
+int OptimizedNvDecoder::Decode(const uint8_t *pData, int nSize, int nFlags, int64_t nTimestamp) {
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h:51]: why not put the function definition into a standalone cpp file?

File: third_party/Video_Codec_SDK/Samples/Utils/Logger.h [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+#include <string>
+#include <sstream>
+#include <mutex>
+#include <time.h>
+
+#ifdef _WIN32
+#include <winsock.h>
+#include <windows.h>
+
+#pragma comment(lib, "ws2_32.lib")
```
Comment (hippogr) [inline third_party/Video_Codec_SDK/Samples/Utils/Logger.h:41]: Sorry I found this is from NVidia code, right? If so, just igore the comments...

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+
+        std::ostringstream errorString;
+        errorString << std::endl
+                    << "MBCount             : " << (pVideoFormat->coded_width >> 4) * (pVideoFormat->coded_height >> 4)
+                    << std::endl
+                    << "Max Supported mbcnt : " << m_decodecaps.nMaxMBCount << std::endl
+                    << "MBCount not supported on this GPU";
+
+        const std::string cErr = errorString.str();
+        NVDEC_THROW_ERROR(cErr, CUDA_ERROR_NOT_SUPPORTED);
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h:181]: a single line like NVDEC_THROW_ERROR(errorString.str(), CUDA_ERROR_NOT_SUPPORTED);

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+    } catch (const std::exception &e) {
+        std::cerr << "Exception in deocding: " << e.what() << std::endl;
+        return 0;
+    }
+}
+
+/**
+ *  @brief  Function to read the video paths from a file
+ */
+std::vector<std::string> ReadMultipleVideoFiles(std::string filepath) {
```
Comment (hippogr) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp:206]: >std::string [](http://example.com/codeflow?start=48&length=11)

const std::string&

Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp]: updated in ReadMultipleVideoFiles(const std::string &filepath)  and GenerateTotalFileList(const std::string &inputFilesListPath, int nTotalVideo,
                                               const char *szInFilePath)

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h (hunk 10 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h]: ok, updated

File: third_party/Video_Codec_SDK/Samples/Utils/Logger.h (hunk 9 above)
Comment (yukirora) [reply third_party/Video_Codec_SDK/Samples/Utils/Logger.h:41]: yes, I click the resolve for all third party directly

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h (hunk 8 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h]: seperated

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h (hunk 7 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h:40]: added

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp [hunk 12]
```diff
@@ -67,15 +67,19 @@ void ShowHelpAndExit(const char *szBadOption = NULL) {
...
-        << "-multi_input The file path which lists the path of multiple video in each line" << std::endl
-        << "-codec       The codec of video to test" << std::endl;
+        << "-i           Input file path. No default value. One of -i and -multi_input is required." << std::endl
+        << "-o           Output file path of raw data. No default value. Optional." << std::endl
+        << "-gpu         Ordinal of GPU to use. Default 0. Optional." << std::endl
+        << "-thread      Number of decoding thread. Default 5. Optional." << std::endl
+        << "-total       Number of total video to test. Default 100. Optional." << std::endl
+        << "-single      (No value) Use single cuda context for every thread. Default is multi-context, one context "
+           "per thread."
+        << std::endl
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp:77]: Do we need to make "per thread." and << std::endl in single line?

**Review by hippogr — APPROVED** [review]
:shipit:

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp (hunk 2 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp:62]: add it into description

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        m_decodecaps.nBitDepthMinus8 = pVideoFormat->bit_depth_luma_minus8;
+
+        CUDA_DRVAPI_CALL(cuCtxPushCurrent(m_cuContext));
+        NVDEC_API_CALL(cuvidGetDecoderCaps(&m_decodecaps));
+        CUDA_DRVAPI_CALL(cuCtxPopCurrent(NULL));
+    }
+
+    if (!m_decodecaps.bIsSupported) {
+        NVDEC_THROW_ERROR("Codec not supported on this GPU", CUDA_ERROR_NOT_SUPPORTED);
+        return nDecodeSurface;
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h]: yes, we have catch in decode function

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h (hunk 1 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/OptimizedNvDecoder.h]: updated

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp (hunk 12 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp:77]: merge 'per thread' into last line but << std::endl is by clang-format

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp [hunk 14]
```diff
@@ -67,15 +67,19 @@ void ShowHelpAndExit(const char *szBadOption = NULL) {
...
+        << "-host        (No value) Copy frame to host memory .Default is "
+           "device memory)"
+        << std::endl
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp]: merge 'device memory' with last line, but <<std::endl is by clang-format

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp [hunk 15]
```diff
@@ -394,21 +410,33 @@ int main(int argc, char **argv) {
...
+        for (int i = 0; i < vnLatency.size(); i++) {
+            if (vnLatency[i] != 0) {
+                videoFPS.push_back(vnFrame[i] / vnLatency[i]);
+            }
+        }
+        std::tie(sum, mean, min, max, p50, p90, p95, p99) = CalMetrics(videoFPS);
+        std::cout << "Mean FPS for each video=" << mean << " P50 FPS=" << p50 << " P90 FPS=" << p90
+                  << " P95 FPS=" << p95 << " P99 FPS=" << p99 << std::endl;
+        std::tie(sum, mean, min, max, p50, p90, p95, p99) = CalMetrics(frLatency);
+        std::cout << "Mean Latency for each frame=" << mean * 1000 << " P50 Latency=" << p50 * 1000
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp:434]: in fact, no, removed LatencyPerFrame, it use total time/ total frames, but it seems not reasonable for multi-thread scenario

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp [hunk 16]
```diff
@@ -394,21 +410,33 @@ int main(int argc, char **argv) {
         std::vector<int> vnFrame(nTotalVideo);
         int nTotalFrames = 0;
         std::vector<double> vnLatency;
-
-        auto elapsedTime = run(vDec, nThread, files, vnFrame, vExceptionPtrs, &nTotalFrames, vnLatency);
+        std::vector<double> frLatency;
+        auto elapsedTime = run(vDec, nThread, files, vnFrame, vExceptionPtrs, &nTotalFrames, vnLatency, frLatency);
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/AppDecPerf.cpp]: yes, merged them into run

**Pull request merged**